tls/s2n_handshake_io.c

```diff
@@ -1314,23 +1279,20 @@ static int s2n_handle_retry_state(struct s2n_connection *conn)
         conn->in_status = ENCRYPTED;
     }
 
-    if (r < 0) {
-        /* There is some other problem and we should kill the connection. */
-        if (conn->session_id_len) {
-            s2n_try_delete_session_cache(conn);
-        }
-
-        POSIX_GUARD(s2n_connection_kill(conn));
-        S2N_ERROR_PRESERVE_ERRNO();
-    }
-
     if (CONNECTION_IS_WRITER(conn)) {
+        POSIX_GUARD(r);
+
         /* If we're the writer and handler just finished, update the record header if
          * needed and let the s2n_handshake_write_io write the data to the socket */
         if (EXPECTED_RECORD_TYPE(conn) == TLS_HANDSHAKE) {
             POSIX_GUARD(s2n_handshake_finish_header(&conn->handshake.io));
         }
     } else {
+        if (r < 0 && conn->session_id_len) {
+            s2n_try_delete_session_cache(conn);
+        }
+        WITH_ERROR_BLINDING(conn, POSIX_GUARD(r));
```

Comment:
https://github.com/aws/s2n-tls/pull/2852#pullrequestreview-667027652

Previously we always called s2n_connection_kill(conn) on error, for both readers and writers. I moved error handling down into the CONNECTION_IS_WRITER if/else so readers and writers do it differently, like in s2n_negotiate.